config-gen/apis/v1alpha1/types.go

```diff
@@ -0,0 +1,192 @@
+	WebhooksComponent = "webhooks"
+
+	// CRDsComponent will create the CRDs
+	CRDsComponent = "crds"
+
+	// RBACComponent will create RBAC rules
+	RBACComponent = "rbac"
+
+	// CertManagerComponent will create the Issuer and Certificate resources
+	// for CertManager to inject the certificates
+	CertManagerComponent = "cert-manager"
+
+	// PrometheusComponent will create the ServiceMonitor resource
+	PrometheusComponent = "prometheus"
+)
+
+// getDefaultComponents returns the set of components that are created by default
+func getDefaultComponents() map[string]bool {
+	return map[string]bool{
+		NamespaceComponent:         true,
+		ControllerManagerComponent: true,
+		RBACComponent:              true,
+		CRDsComponent:              true,
+	}
+}
+
+// Enabled returns true if the component is enabled
+func (a *KubebuilderProjectSpec) Enabled(component string) bool {
+	return a.Components[component]
+}
+
+// ComponentConfigEnabled returns true if component config is being used
+func (a *KubebuilderProjectSpec) ComponentConfigEnabled() bool {
+	return a.ComponentConfigFilepath != ""
+}
+
+// KubebuilderProjectSpec defines the desired configuration to be generated
+type KubebuilderProjectSpec struct {
+	// Directory is the kubebuilder directory containing the code.
+	// Defaults to the directoring containing the KubebuilderProject configuration file.
+	Directory string `json:"projectDirectory" yaml:"projectDirectory"`
+
+	// Name is the name of project and used to generate the component and role names
+	// Defaults to metadata.name
+	Name string `json:"projectName" yaml:"projectName"`
+
+	// Namespace is the namespace to run the project in.
+	// Defaults to spec.name-system
+	Namespace string `json:"namespace" yaml:"namespace"`
+
+	// Image is the container image to run in the controller-manager Deployment
+	Image string `json:"image" yaml:"image"`
+
+	// Components is a map of components to enable or disable.
+	// Defaults to enabling:  ["namespace", "controller-manager", "rbac", "crds"]
+	Components map[string]bool `json:"components" yaml:"components"`
+
+	// DisableAuthProxy if set to true will disable the auth proxy
+	DisableAuthProxy bool `json:"disableAuthProxy,omitempty" yaml:"disableAuthProxy,omitempty"`
+
+	// ConversionWebhooks is a map of kinds to enable conversion webhooks for
+	ConversionWebhooks map[string]bool `json:"conversionWebhooks,omitempty" yaml:"conversionWebhooks,omitempty"`
+
+	// Development contains development options
+	Development DevelopmentOptions `json:"developmentOptions,omitempty" yaml:"developmentOptions,omitempty"`
+
+	// ComponentConfigFilepath is the relative path to a component config file used to configure the
+	// controller-manager
+	ComponentConfigFilepath string `json:"componentConfigFilepath,omitempty" yaml:"componentConfigFilepath,omitempty"`
+}
+
+// KubebuilderProjectStatus is runtime status for the api configuration
+type KubebuilderProjectStatus struct {
+	CertCA string
+
+	CertKey string
```

Comment:
Why CertCa and CertKey are attrs for the status options and not spec?

Reply:
Because they shouldn't be specified by the user.

Reply:
So, have we not any scenario where people would like to use their own CertCA and CertKey? 
And then, should it always be provided by the config gen?

Reply:
I think we should think carefully before exposing these as an API.  Exposing these directly seems like a poor use experience, rather instead I would imagine the user specify the path to a .pem file or something.  Or using kustomize to generate the secret from a file + annotating it, and pulling that in automatically.  There are a number of ways we may want to do this, but I don't think manually specifying these fields is one of them.

Reply:
+1 for your thoughts. However, `KubebuilderConfigGenStatus` shows only be useful to Cert/Key features. So, I still not sure about its usage as a status. we can re-think that latter as well. Not a blocker for this PR anyway.

Reply:
SG.  The status exists as a way of making data available to the go-templates at runtime.